### PR TITLE
Pr2022 34/auth services

### PR DESCRIPTION
**Pull request by KetiRudneva**
Using Axios interceptor request for API token

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,7 +22,8 @@
     "react-query": "3.9.4",
     "react-router": "5.2.1",
     "react-router-dom": "5.3.0",
-    "recoil": "0.6.1"
+    "recoil": "0.6.1",
+    "yarn": "^1.22.17"
```
Comment (Neoclassic-alt) [inline package.json:26]: Don't use fixed versions (with ^)

File: src/data/auth/authServices.ts [hunk 2]
```diff
@@ -1,3 +1,52 @@
...
+  SignInUserData,
+  SignInUserResponse,
+  RefreshTokenData,
+  RefreshTokenResponse,
+  LogoutUserResponse,
+} from '../../types/auth/auth';
+import axiosInstance from '../http/http';
 import { signInUserMock } from './__mocks/authServicesMocks';
 
 export const signInUser = signInUserMock;
```
Comment (Neoclassic-alt) [inline src/data/auth/authServices.ts:13]: If mocks methods don't use, must delete it

File: src/data/auth/userServices.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import { getCurrentUserMock } from '../user/__mocks/userServicesMocks';
+
+export const getCurrentUser = getCurrentUserMock;
```
Comment (Neoclassic-alt) [inline src/data/auth/userServices.ts:3]: Similarly

File: src/modules/Auth/hooks/useSignUpUser.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { useSignUnUserMutation } from '../../../queries/auth';
```
Comment (Neoclassic-alt) [inline src/modules/Auth/hooks/useSignUpUser.ts:1]: Preferably write symbol @ instead "../../../"

File: package.json (hunk 1 above)
Comment (NKonstantin007) [reply package.json:26]: We needn't the `yarn` as a dependency of project. You need to remove this line.

File: src/data/auth/authServices.ts [hunk 5]
```diff
@@ -1,3 +1,52 @@
...
+
+export const signUpUser = (
+  signUpUserData: SignUpUserData,
+): Promise<SignUnUserResponse> => {
+  return axiosInstance
+    .post<SignUnUserResponse>(`signup`, signUpUserData)
+    .then((res) => {
+      return res.data;
+    });
+};
```
Comment (NKonstantin007) [inline src/data/auth/authServices.ts:23]: I prefer to use async/await syntax. It help to organize the code and make code more clarify and structural. Can you please update the all services using this syntax?
Example:
```suggestion
export const signUpUser = async (
  signUpUserData: SignUpUserData,
): Promise<SignUnUserResponse> => {
  const { data } =  await axiosInstance.post<SignUnUserResponse>(`signup`, signUpUserData);
   return data;
};
```

File: src/data/http/http.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import axios from 'axios';
+import { refreshToken as refreshTokenService } from '../auth/authServices';
+
+const axiosInstance = axios.create({
+  baseURL: 'http://173.212.214.70:3001/auth/',
```
Comment (NKonstantin007) [inline src/data/http/http.ts:5]: The baseUrl should be `http://173.212.214.70:3001`. And won't forget to add `/auth/` segment to the pathes of the auth services

File: src/data/http/http.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import axios from 'axios';
+import { refreshToken as refreshTokenService } from '../auth/authServices';
+
+const axiosInstance = axios.create({
```
Comment (NKonstantin007) [inline src/data/http/http.ts:4]: I think `http` is a good name for this file. And It seems that http is more suitable name then axiosInstance.
I suggest to rename:
```suggestion
const http = axios.create({
```

File: src/data/http/http.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      return Promise.reject(error);
+    }
+    const res = await refreshTokenService({ refreshToken });
+    localStorage.setItem('token', res.token);
+    localStorage.setItem('refreshToken', res.refreshToken);
+    return axiosInstance(error.config);
+  },
+);
+
+export default axiosInstance;
```
Comment (NKonstantin007) [inline src/data/http/http.ts:42]: I don't like the default export because in the files where you need to import this module you need to think how to name the module correctly. And when you export some entities not using default export you always known the name of module.
So I suggest to export axiosInstance not using default export

File: src/data/user/userServices.ts [hunk 9]
```diff
@@ -1,3 +0,0 @@
-import { getCurrentUserMock } from './__mocks/userServicesMocks';
```
Comment (NKonstantin007) [inline src/data/user/userServices.ts:1]: Why did you delete this file? I don't see the reason to delete it

File: src/data/user/index.ts [hunk 10]
```diff
@@ -1 +1 @@
-export * from './userServices';
+export * from '../auth/userServices';
```
Comment (NKonstantin007) [inline src/data/user/index.ts:1]: Return userServices.ts file to the `data/user` folder. It shouldn't be in `auth` filder

File: src/types/auth/auth.ts [hunk 11]
```diff
@@ -1,8 +1,31 @@
...
   accessToken: string;
 };
+
+export type SignUpUserData = {
+  email: string;
+  name: string;
+  password: string;
+};
+
+export type SignUnUserResponse = {
```
Comment (NKonstantin007) [inline src/types/auth/auth.ts:16]: Typo
```suggestion
export type SignUpUserResponse = {
```

File: src/queries/auth/authMutations.ts [hunk 12]
```diff
@@ -1,9 +1,40 @@
...
   return useMutation('sign-in-user-mutation', signInUser, options);
 }
+
+export function useSignUnUserMutation(
+  options?: UseMutationOptions<SignUnUserResponse, Error, SignUpUserData>,
+) {
+  return useMutation('sign-in-user-mutation', signUpUser, options);
+}
+
+export function useRefreshTokenMutation(
```
Comment (NKonstantin007) [inline src/queries/auth/authMutations.ts:30]: You can delete this mutation.Since service `refreshToken` won't be called in some part of app except a response interceptor

**Review by NKonstantin007 — CHANGES_REQUESTED** [review]

File: src/data/auth/authServices.ts [hunk 13]
```diff
@@ -1,3 +1,52 @@
...
 
 export const signInUser = signInUserMock;
+
+export const signUpUser = (
+  signUpUserData: SignUpUserData,
+): Promise<SignUnUserResponse> => {
+  return axiosInstance
+    .post<SignUnUserResponse>(`signup`, signUpUserData)
+    .then((res) => {
+      return res.data;
```
Comment (NKonstantin007) [inline src/data/auth/authServices.ts:21]: Please  create a ticket is Jira for improvements: Create and http methods which will returns  response.data. It will be low priority ticket.
It will allow to do like this:

```typescript
export const signUpUser = (
  signUpUserData: SignUpUserData,
): Promise<SignUnUserResponse> => {
  return await post<SignUnUserResponse>(`signup`, signUpUserData);
```

File: src/data/http/http.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  baseURL: 'http://173.212.214.70:3001/auth/',
+});
+
+axiosInstance.interceptors.request.use(
+  (config) => {
+    return {
+      ...config,
+      headers: {
+        ...config.headers,
+        Authorization: `Baerer ${localStorage.getItem('token')}`,
```
Comment (NKonstantin007) [inline src/data/http/http.ts:14]:
```suggestion
        Authorization: `Bearer ${localStorage.getItem('token')}`,
```

File: src/data/http/http.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+axiosInstance.interceptors.response.use(
+  (config) => {
+    return config;
+  },
+  async (error) => {
+    if (!axios.isAxiosError(error)) {
+      return Promise.reject(error);
+    }
+    const refreshToken = localStorage.getItem('refreshToken');
+    if (!refreshToken || error.response?.status === 401) {
```
Comment (NKonstantin007) [inline src/data/http/http.ts:32]:
```suggestion
    if (!refreshToken || error.response?.status !== 401) {
```

**Review by NKonstantin007 — CHANGES_REQUESTED** [review]

File: src/data/auth/authServices.ts [hunk 16]
```diff
@@ -1,3 +1,52 @@
...
+  return axiosInstance
+    .post<RefreshTokenResponse>(`refresh`, refreshTokenData)
+    .then((res) => {
+      return res.data;
+    });
+};
+
+export const logoutUser = (): Promise<LogoutUserResponse> => {
+  return axiosInstance.post<LogoutUserResponse>(`logout`).then((res) => {
+    localStorage.removeItem('user');
```
Comment (NKonstantin007) [inline src/data/auth/authServices.ts:49]:
```suggestion
    localStorage.removeItem('token');
    localStorage.removeItem('refreshToken');
```

**Review by NKonstantin007 — CHANGES_REQUESTED** [review]

File: src/data/user/userServices.ts (hunk 9 above)
Comment (KetiRudneva) [reply src/data/user/userServices.ts:1]: My project has this file. It's weird.

File: src/data/auth/userServices.ts (hunk 3 above)
Comment (NKonstantin007) [inline src/data/auth/userServices.ts:3]: Delete this file

File: src/data/http/http.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+      return Promise.reject(error);
+    }
+    const refreshToken = localStorage.getItem('refreshToken');
+    if (!refreshToken || error.response?.status !== 401) {
+      return Promise.reject(error);
+    }
+    const res = await refreshTokenService({ refreshToken });
+    localStorage.setItem('token', res.token);
+    localStorage.setItem('refreshToken', res.refreshToken);
+    return http(error.config);
```
Comment (NKonstantin007) [inline src/data/http/http.ts:38]: Here is asyncronous request therefore we need to wrap in try/catch
```suggestion
  try {
    const res = await refreshTokenService({ refreshToken });
    localStorage.setItem('token', res.token);
    localStorage.setItem('refreshToken', res.refreshToken);
    return http(error.config);
  catch (refreshError) {
  	return return Promise.reject(refreshError);
  }
```

File: src/queries/auth/authMutations.ts [hunk 18]
```diff
@@ -1,9 +1,27 @@
...
+} from '../../types/auth';
 
 export function useSignInUserMutation(
   options?: UseMutationOptions<SignInUserResponse, Error, SignInUserData>,
 ) {
-  return useMutation('sign-in-user-mutation', signInUser, options);
+  return useMutation('sign-in-user-mutation', loginUser, options);
+}
+
+export function useSignUnUserMutation(
```
Comment (NKonstantin007) [inline src/queries/auth/authMutations.ts:17]: I didn't notice the typo during the first review
```suggestion
export function useSignUpUserMutation(
```

File: src/queries/auth/authMutations.ts [hunk 19]
```diff
@@ -1,9 +1,27 @@
...
+  options?: UseMutationOptions<SignUpUserResponse, Error, SignUpUserData>,
+) {
+  return useMutation('sign-in-user-mutation', signUpUser, options);
```
Comment (NKonstantin007) [inline src/queries/auth/authMutations.ts:20]:
```suggestion
  return useMutation('sign-up-user-mutation', signUpUser, options);
```

File: src/queries/auth/authMutations.ts [hunk 20]
```diff
@@ -1,9 +1,27 @@
...
+}
+
+export function useLogoutUserMutation(
+  options?: UseMutationOptions<LogoutUserResponse, Error>,
+) {
+  return useMutation('sign-in-user-mutation', logoutUser, options);
```
Comment (NKonstantin007) [inline src/queries/auth/authMutations.ts:26]:
```suggestion
  return useMutation('logout-user-mutation', logoutUser, options);
```

**Review by NKonstantin007 — CHANGES_REQUESTED** [review]

File: src/data/auth/userServices.ts (hunk 3 above)
Comment (KetiRudneva) [reply src/data/auth/userServices.ts:3]: You said return this file in folder 2 days ago

**Review by NKonstantin007 — CHANGES_REQUESTED** [review]
